### DISCUSSION TITLE
Exceptions raised through HomeAssistantError and subclasses are expected and stacktraces are not warranted

### DISCUSSION THREAD

**Discussion by jbouwh** [home-assistant/architecture#986] [General] [answered]
## Background

We use `HomeAssistantError` and sub classes of it (e.g. `ConfigError`), to re-raise expected exceptions that were caught and also to signal expected errors. We attach a user friendly message to the exception to allow users to see the exception in the UI.
The messages should be readable and make the cause clear, and a stack trace is not needed.

However, the stack trace of the original exception can still be of interest as a development aid, in particular for debugging purposes.

## Problem

There's currently no solution which satisfies both the desire to not spam user's logs with unwanted stack traces while at the same time giving developers stack traces which aid development. This sometimes leads to a series of PR where stack traces are removed, only to be readded later and so on.

## Proposal

Suppress stack traces for `HomeAssistantError` and its subclasses in logs unless either of the following is true:
- The core build is not a stable build
- The logger printing the `HomeAssistantError` has its level set to DEBUG

This proposal is implemented in core PR [#101762](https://github.com/home-assistant/core/pull/101762).

## Concerns

With this suggestion, a `HomeAssistantError` which is never caught will also not get a stack trace. If needed, we can add a mechanism for making sure stack traces are always printed in such cases, for example by setting an attribute on the `HomeAssistantError` or by making certain loggers always print stack traces.

**Comment by elupus** [comment, marked as answer]
This can lead to memleaks. If an exception is unexpected (nobody catching it). We can leak file handles, forget to finish futures. Just a out anything. We camt hide the backtrace for such a thing.

It does not matter its a user facing exception message. Its still a bug if nobody catches it.

In essence, the decision as to if an exception (of  any sort since its a stack unwind) is expected or not, can only be made when its caught.

If we catch a HomeAssistantError at a known location where we know the callstack likely handled it properly, we should hide all callstacks from user (unless debug logs are on)

If it bubbles out to some generic catch all exception, it is obviously not an error we expected a d it should be logged in all its gory glory so we actually fix the bug it is when the exception bubbles to that location. Fixing could be as simple as catching the HomeAssistantError and just logging the message.

**Reply by jbouwh** [reply]
`HomeAssistantError` is raised for Service Calls and catched explictly to be printed in the UI.
To do that these errors on service calls are catched here:
- https://github.com/home-assistant/core/blob/857f2e1d869302c4b8f871dcdedd514ed1350ded/homeassistant/components/websocket_api/commands.py#L210-L244.

So do you mean the concern is about cases exceptions (based on `HomeAssistantError`) are raised that are not part of a service call?

**Reply by elupus** [reply]
Yes those are probably mostly bugs. Say this:

```py
@callback
def async_get_hass() -> HomeAssistant:
    """Return the HomeAssistant instance.

    Raises HomeAssistantError when called from the wrong thread.

    This should be used where it's very cumbersome or downright impossible to pass
    hass to the code which needs it.
    """
    if not _hass.hass:
        raise HomeAssistantError("async_get_hass called from the wrong thread")
    return _hass.hass
```

This is a huge bug if we ever trigger. If we hide that its very very wrong.

**Reply by jbouwh** [reply]
Follow up discussion:
https://github.com/home-assistant/architecture/discussions/992
